feat: support multiple sources for a single instrument name

PouyaMohseni opened the pull request:
Introduces a multi-source data model for instrument names:

- Added new Source and InstrumentNameSource models
- Added uniqueness constraint enforcing a single main source per instrument name
- Updated import pipeline to store Wikidata as a Source and create link objects
- Updated detail templates to display main source and its verification status
- Updated name creation and deletion workflows for UMIL DB
- Updated views
- Updated admin interfaces to support source links and reviewer permissions

resolved #436

dchiller reviewed (commented):
Two data modeling-related questions:
1. Can you explain more about the decision to model the Instrument Name - Source relationship as a many-to-many field (as opposed to, in particular, a foreign key from Source to Instrument Name), with attributes like contributor and verification status adhering to the relationship table rather than to the name or to the source itself? A few things that come up:
a. We’re introducing here some concept of identity of a “source” that we don’t have any other machinery to manage and/or enforce (and I’m not sure that we necessarily want one). Multiple names might be sourced from Wikidata, and in this PR you make use of a single source with the name “Wikidata” — this makes sense. But what else counts as a unique source? User-generated sources are just text fields — how do we disambiguate or link sources that are provided free-form? Is it important to do that? 
b. Is the name-source relationship really what we are verifying? Or are we verifying a name *based on* some source(s)? Same goes for the other categories of verification: what does it mean for a name - source relationship to be rejected. 
2. What is the role of the `is_main` field? Is it simply a way to set whether a source is displayed (this is what the last paragraph of #436 suggests to me…)? Or does it have additional meaning (for example, is it connected to whether a source is verifiable or not? does it have something to do with “source quality”) If the former, a simple flag makes sense (although perhaps a clearer name would be something like `is_display`). If the latter, there is probably a name that would be more meaningful than “main”

dchiller commented on web-app/django/VIM/apps/instruments/views/update_umil_db.py:
```diff
@@ -148,8 +180,11 @@ def delete_name(request: HttpRequest) -> JsonResponse:
+    # If user is a superuser or contributor to the main source, allow deletion
+    if (
+        request.user.is_superuser
+        or instrument_name.main_source.contributor == request.user
```
> I believe the deletion workflow would be better implemented if we introduce a Contributor model and link it to the InstrumentNameSource model. This is more appropriate because the verification status is not bond to any contributor of a source. If a source has multiple contributors, it only needs to be verified once by a reviewer, regardless of how many contributors it has.

What differentiates a `Contributor` model from the `User` model?

> The InstrumentNameSource relation should be deleted only when all contributors associated with that relation have deleted it. This creates a hierarchy: an instrument name is deleted only when all of its sources are deleted, and a source is deleted only when all of its contributors remove it.

I think the problem here would be that it then becomes hard to communicate to the user what deleting a name means, and what effect the "delete" action should have. When I delete something in Wikidata, it is clear and obvious what the result of that action was: it now no longer shows up on the page. If we want to give users some sort of delete permission, how do we communicate the effect of that with the procedure you're suggesting.

PouyaMohseni committed:
feat(models)!: support multiple sources for a single instrument name

- Add new `Source` and `InstrumentNameSource` models
- Many-to-many relationship between InstrumentName and Source
- Move contributor status onto the relationship table
- Update name creation and deletion workflows to operate on relationships
- Update Wikidata import to create Source objects and link Name–Source relations
- Update admin, views, and templates to display main source and status
- Remove restriction on adding new sources after verification

Resolves  #436

PouyaMohseni committed:
fix: address minor issues from review

- used `iterator()` in admin.py
- changed `help_text` of contributor

ahankinson commented:
How many queries does this functionality generate? Have you benchmarked it with Django Debug Toolbar and checked?

PouyaMohseni committed:
refator: use `prefetch_related()` for prefetching sources to reduce the dataset request hits
